Source/SwiftLintFramework/Rules/Lint/EmptyXCTestMethodRule.swift

```diff
@@ -14,34 +14,64 @@ public struct EmptyXCTestMethodRule: Rule, OptInRule, ConfigurationProviderRule
         triggeringExamples: EmptyXCTestMethodRuleExamples.triggeringExamples
     )
 
-    public func validate(file: SwiftLintFile) -> [StyleViolation] {
-        return testClasses(in: file).flatMap { violations(in: file, for: $0) }
+    public func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor? {
+        EmptyXCTestMethodRuleVisitor()
+    }
+}
+
+private class EmptyXCTestMethodRuleVisitor: SyntaxVisitor, ViolationsSyntaxVisitor {
+    private(set) var violationPositions: [AbsolutePosition] = []
+
+    override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+        guard let inheritanceList = node.inheritanceClause?.inheritedTypeCollection else {
+            return .skipChildren
+        }
+        let isTestClass = inheritanceList
+            .map(\.typeName)
+            .compactMap { $0.as(SimpleTypeIdentifierSyntax.self) }
+            .map(\.name)
+            .map(\.text)
+            .contains("XCTestCase")
+        return isTestClass ? .visitChildren : .skipChildren
+    }
+
+    override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
+        .skipChildren
+    }
+
+    override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+        .skipChildren
+    }
+
+    override func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
+        .skipChildren
     }
 
-    // MARK: - Private
+    override func visitPost(_ node: FunctionDeclSyntax) {
+        if node.isOverride, node.hasEmptyBody {
+            violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)
+        } else if node.isTestMethod, node.hasEmptyBody {
+            violationPositions.append(node.funcKeyword.positionAfterSkippingLeadingTrivia)
+        }
+    }
+}
 
-    private func testClasses(in file: SwiftLintFile) -> [SourceKittenDictionary] {
-        let dict = file.structureDictionary
-        return dict.substructure.filter { dictionary in
-            guard dictionary.declarationKind == .class else { return false }
-            return dictionary.inheritedTypes.contains("XCTestCase")
+private extension FunctionDeclSyntax {
+    var isOverride: Bool {
+        if let modifiers = modifiers {
+            return modifiers.map(\.name).map(\.text).contains("override")
```

Comment:
Same here regarding not iterating multiple times and stopping iteration as soon as a match is found.

```suggestion
            return modifiers.contains { $0.name.text == "override" }
```